### PR TITLE
inductor: fix mm_plus_mm fusion pattern issue when has broadcast add

### LINKED ISSUE #100670: [torch.compile] RuntimeError after `mm_plus_mm` fusion
### 🐛 Describe the bug

`torch.compile` will return a RuntimeError below after `mm_plus_mm` fusion

```
extern_kernels.addmm(buf4, buf5, buf6, alpha=1, beta=1, out=buf7)
RuntimeError: The expanded size of the tensor (1) must match the existing size (2) at non-singleton dimension 1.  Target sizes: [1, 1].  Tensor sizes: [1, 2]
```

Here is the reproduced code

```py
import torch

torch.manual_seed(420)

x = torch.randn(1, 3)

class Model(torch.nn.Module):

    def __init__(self):
        super(Model, self).__init__()
        self.fc1 = torch.nn.Linear(3, 4)
        self.fc2 = torch.nn.Linear(4, 2)

    def forward(self, x):
        x = self.fc1(x)
        y = torch.mm(x, torch.ones((4, 2)))
        x = self.fc2(x)
        z = torch.mm(x, torch.ones((2, 1)))
        out = torch.add(y, z)
        return out

func = Model().to('cpu')

with torch.no_grad():
    func.train(False)
    res1 = func(x) # without jit
    print(res1)

    jit_func = torch.compile(func)
    res2 = jit_func(x)
    print(res2)
    # extern_kernels.addmm(buf4, buf5, buf6, alpha=1, beta=1, out=buf7)
    # RuntimeError: The expanded size of the tensor (1) must match the existing size (2) at non-singleton dimension 1.  Target sizes: [1, 1].  Tensor sizes: [1, 2]
```

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230502+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.9.16 (main, Mar  8 2023, 14:00:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.19.5-051905-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.5.119
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 510.108.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          24
On-line CPU(s) list:             0-23
Vendor ID:                       GenuineIntel
Model name:                      12th Gen Intel(R) Core(TM) i9-12900K
CPU family:                      6
Model:                           151
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        2
CPU max MHz:                     6700.0000
CPU min MHz:                     800.0000
BogoMIPS:                        6374.40
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       640 KiB (16 instances)
L1i cache:                       768 KiB (16 instances)
L2 cache:                        14 MiB (10 instances)
L3 cache:                        30 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.2
[pip3] pytorch-triton==2.1.0+7d1a95b046
[pip3] torch==2.1.0.dev20230502+cu118
[pip3] torchaudio==2.1.0.dev20230502+cu118
[pip3] torchvision==0.16.0.dev20230502+cu118
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+7d1a95b046          pypi_0    pypi
[conda] torch                     2.1.0.dev20230502+cu118          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230502+cu118          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230502+cu118          pypi_0    pypi
```

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100679

cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/kernel/mm_plus_mm.py [hunk 1]
```diff
@@ -151,7 +151,7 @@ def tuned_mm_plus_mm(mat1, mat2, mat3, mat4, *, layout=None):
     ):
         # TODO(jansel): support different K values when this is fixed:
         # https://github.com/openai/triton/issues/967
-        return lowerings[aten.addmm](lowerings[aten.mm](mat1, mat2), mat3, mat4)
+        return lowerings[aten.addmm](lowerings[aten.mm](mat3, mat4), mat1, mat2)
```
Comment (jgong5) [inline torch/_inductor/kernel/mm_plus_mm.py:154]: Can this work when `mat1@mat2` needs broadcasting, not `mat3@mat4`?

Comment (ngimel) [inline torch/_inductor/kernel/mm_plus_mm.py:154]: this doesn't seem like a correct fix, as arguments could come in a different order, and mm(mat1, mat2) would be a small matrix that would need to be broadcasted (and will fail to be broadcasted) to mm(mat3, mat4). In this case we need to return the original expression, mm + mm, and not use addmm (there's a similar fix in a pattern that matches add + mm  to addmm, sometimes it's impossible)

Comment (XiaobingSuper) [reply torch/_inductor/kernel/mm_plus_mm.py]: Yes, the original is not a correct fix, please see the latest code.

Comment (XiaobingSuper) [reply torch/_inductor/kernel/mm_plus_mm.py]: The code is updated.

File: torch/_inductor/kernel/mm_plus_mm.py [hunk 2]
```diff
@@ -151,25 +153,26 @@ def tuned_mm_plus_mm(mat1, mat2, mat3, mat4, *, layout=None):
...
+            return lowerings[aten.addmm](lowerings[aten.mm](mat3, mat4), mat1, mat2)
+        return lowerings[aten.add](
+            lowerings[aten.mm](mat1, mat2), lowerings[aten.mm](mat3, mat4)
+        )
 
     # options to tune from
-    choices = [aten_mm_plus_mm.bind((mat1, mat2, mat3, mat4), layout)]
-    if use_triton_template(layout):
+    choices = [aten_mm_plus_mm.bind((mat1, mat2, mat3, mat4), layout1)]
+    if use_triton_template(layout1):
```
Comment (jgong5) [inline torch/_inductor/kernel/mm_plus_mm.py:164]: assert `layout1 == layout2` here?

File: torch/_inductor/kernel/mm_plus_mm.py [hunk 3]
```diff
@@ -151,25 +153,26 @@ def tuned_mm_plus_mm(mat1, mat2, mat3, mat4, *, layout=None):
     ):
         # TODO(jansel): support different K values when this is fixed:
         # https://github.com/openai/triton/issues/967
-        return lowerings[aten.addmm](lowerings[aten.mm](mat1, mat2), mat3, mat4)
-
-    m, n, k, layout, mat1, mat2 = mm_args(mat1, mat2, layout=layout)
-    m, n, k, layout, mat3, mat4 = mm_args(mat3, mat4, layout=layout)
+        if m1 == m2 and n1 == n2:
```
Comment (jgong5) [inline torch/_inductor/kernel/mm_plus_mm.py:156]: I guess we need guards here?

**Review by ngimel — APPROVED** [review]

**Review by jansel — APPROVED** [review]

File: torch/_inductor/kernel/mm_plus_mm.py (hunk 2 above)
Comment (XiaobingSuper) [reply torch/_inductor/kernel/mm_plus_mm.py:167]: Yes, added.

File: torch/_inductor/kernel/mm_plus_mm.py (hunk 3 above)
Comment (XiaobingSuper) [reply torch/_inductor/kernel/mm_plus_mm.py:156]: added.

**Review by jgong5 — APPROVED** [review]

**Pull request closed without merging**